{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unsupervised Learning with Gaussian Processes\n",
    "=============================================\n",
    "\n",
    "### [Neil D. Lawrence](http://inverseprobability.com), University of\n",
    "\n",
    "Cambridge"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Abstract**: null"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "$$\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- Do not edit this file locally. -->\n",
    "<!-- Do not edit this file locally. -->\n",
    "<!---->\n",
    "<!-- Do not edit this file locally. -->\n",
    "<!-- Do not edit this file locally. -->\n",
    "<!-- The last names to be defined. Should be defined entirely in terms of macros from above-->\n",
    "<!--\n",
    "\n",
    "-->"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Setup\n",
    "-----\n",
    "\n",
    "First we download some libraries and files to support the notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "urllib.request.urlretrieve('https://raw.githubusercontent.com/lawrennd/talks/gh-pages/mlai.py','mlai.py')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "urllib.request.urlretrieve('https://raw.githubusercontent.com/lawrennd/talks/gh-pages/teaching_plots.py','teaching_plots.py')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "urllib.request.urlretrieve('https://raw.githubusercontent.com/lawrennd/talks/gh-pages/gp_tutorial.py','gp_tutorial.py')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "plt.rcParams.update({'font.size': 22})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!--setupplotcode{import seaborn as sns\n",
    "sns.set_style('darkgrid')\n",
    "sns.set_context('paper')\n",
    "sns.set_palette('colorblind')}-->"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "pods\n",
    "----\n",
    "\n",
    "In Sheffield we created a suite of software tools for ‘Open Data\n",
    "Science’. Open data science is an approach to sharing code, models and\n",
    "data that should make it easier for companies, health professionals and\n",
    "scientists to gain access to data science techniques.\n",
    "\n",
    "You can also check this blog post on [Open Data\n",
    "Science](http://inverseprobability.com/2014/07/01/open-data-science).\n",
    "\n",
    "The software can be installed using"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%pip install --upgrade git+https://github.com/sods/ods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "from the command prompt where you can access your python installation.\n",
    "\n",
    "The code is also available on github:\n",
    "<a href=\"https://github.com/sods/ods\" class=\"uri\">https://github.com/sods/ods</a>\n",
    "\n",
    "Once `pods` is installed, it can be imported in the usual manner."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GPy: A Gaussian Process Framework in Python\n",
    "-------------------------------------------\n",
    "\n",
    "Gaussian processes are a flexible tool for non-parametric analysis with\n",
    "uncertainty. The GPy software was started in Sheffield to provide a easy\n",
    "to use interface to GPs. One which allowed the user to focus on the\n",
    "modelling rather than the mathematics.\n",
    "\n",
    "<img class=\"\" src=\"http://inverseprobability.com/talks/slides/../slides/diagrams/gp/gpy.png\" style=\"width:70%\">\n",
    "\n",
    "Figure: <i>GPy is a BSD licensed software code base for implementing\n",
    "Gaussian process models in Python. It is designed for teaching and\n",
    "modelling. We welcome contributions which can be made through the Github\n",
    "repository\n",
    "<a href=\"https://github.com/SheffieldML/GPy\" class=\"uri\">https://github.com/SheffieldML/GPy</a></i>\n",
    "\n",
    "GPy is a BSD licensed software code base for implementing Gaussian\n",
    "process models in python. This allows GPs to be combined with a wide\n",
    "variety of software libraries.\n",
    "\n",
    "The software itself is available on\n",
    "[GitHub](https://github.com/SheffieldML/GPy) and the team welcomes\n",
    "contributions.\n",
    "\n",
    "The aim for GPy is to be a probabilistic-style programming language,\n",
    "i.e. you specify the model rather than the algorithm. As well as a large\n",
    "range of covariance functions the software allows for non-Gaussian\n",
    "likelihoods, multivariate outputs, dimensionality reduction and\n",
    "approximations for larger data sets.\n",
    "\n",
    "The documentation for GPy can be found\n",
    "[here](https://gpy.readthedocs.io/en/latest/)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Probabilistic PCA\n",
    "=================\n",
    "\n",
    "In 1997 [Tipping and\n",
    "Bishop](http://research.microsoft.com/pubs/67218/bishop-ppca-jrss.pdf)\n",
    "(Tipping and Bishop, 1999) and\n",
    "[Roweis](https://www.cs.nyu.edu/~roweis/papers/empca.pdf) (Roweis, n.d.)\n",
    "independently revisited Hotelling’s model and considered the case where\n",
    "the noise variance was finite, but *shared* across all output dimensons.\n",
    "Their model can be thought of as a factor analysis where $$\n",
    "\\boldsymbol{\\Sigma} = \\noiseStd^2 \\mathbf{I}.\n",
    "$$ This leads to a marginal likelihood of the form $$\n",
    "p(\\mathbf{Y}|\\mathbf{W}, \\noiseStd^2)\n",
    "= \\prod_{i=1}^n\\mathcal{N}\\left(\\mathbf{ y}_{i, :}|\\mathbf{0},\\mathbf{W}\\mathbf{W}^\\top + \\noiseStd^2 \\mathbf{I}\\right)\n",
    "$$ where the limit of $\\noiseStd^2\\rightarrow 0$ is *not* taken. This\n",
    "defines a proper probabilistic model. Tippping and Bishop then went on\n",
    "to prove that the *maximum likelihood* solution of this model with\n",
    "respect to $\\mathbf{W}$ is given by an eigenvalue problem. In the\n",
    "probabilistic PCA case the eigenvalues and eigenvectors are given as\n",
    "follows. $$\n",
    "\\mathbf{W}= \\mathbf{U}\\mathbf{L} \\mathbf{R}^\\top\n",
    "$$ where $\\mathbf{U}$ is the eigenvectors of the empirical covariance\n",
    "matrix $$\n",
    "\\mathbf{S} = \\sum_{i=1}^n(\\mathbf{ y}_{i, :} - \\boldsymbol{ \\mu})(\\mathbf{ y}_{i,\n",
    ":} - \\boldsymbol{ \\mu})^\\top,\n",
    "$$ which can be written\n",
    "$\\mathbf{S} = \\frac{1}{n} \\mathbf{Y}^\\top\\mathbf{Y}$ if the data is zero\n",
    "mean. The matrix $\\mathbf{L}$ is diagonal and is dependent on the\n",
    "*eigenvalues* of $\\mathbf{S}$, $\\boldsymbol{\\Lambda}$. If the $i$th\n",
    "diagonal element of this matrix is given by $\\lambda_i$ then the\n",
    "corresponding element of $\\mathbf{L}$ is $$\n",
    "\\ell_i = \\sqrt{\\lambda_i - \\noiseStd^2}\n",
    "$$ where $\\noiseStd^2$ is the noise variance. Note that if $\\noiseStd^2$\n",
    "is larger than any particular eigenvalue, then that eigenvalue (along\n",
    "with its corresponding eigenvector) is *discarded* from the solution."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Python Implementation of Probabilistic PCA\n",
    "------------------------------------------\n",
    "\n",
    "We will now implement this algorithm in python."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# probabilistic PCA algorithm\n",
    "def ppca(Y, q):\n",
    "    # remove mean\n",
    "    Y_cent = Y - Y.mean(0)\n",
    "\n",
    "    # Comute covariance\n",
    "    S = np.dot(Y_cent.T, Y_cent)/Y.shape[0]\n",
    "    lambd, U = np.linalg.eig(S)\n",
    "\n",
    "    # Choose number of eigenvectors\n",
    "    sigma2 = np.sum(lambd[q:])/(Y.shape[1]-q)\n",
    "    l = np.sqrt(lambd[:q]-sigma2)\n",
    "    W = U[:, :q]*l[None, :]\n",
    "    return W, sigma2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In practice we may not wish to compute the eigenvectors of the\n",
    "covariance matrix directly. This is because it requires us to estimate\n",
    "the covariance, which involves a sum of squares term, before estimating\n",
    "the eigenvectors. We can estimate the eigenvectors directly either\n",
    "through [QR\n",
    "decomposition](http://en.wikipedia.org/wiki/QR_decomposition) or\n",
    "[singular value\n",
    "decomposition](http://en.wikipedia.org/wiki/Singular_value_decomposition).\n",
    "We saw a similar issue arise when , where we also wished to avoid\n",
    "computation of $\\mathbf{Z}^\\top\\mathbf{Z}$ (or in the case of\n",
    "$\\boldsymbol{\\Phi}^\\top\\boldsymbol{\\Phi}$).\n",
    "\n",
    "::: {.cell .markdown}\n",
    "\n",
    "Posterior for Principal Component Analysis\n",
    "==========================================\n",
    "\n",
    "Under the latent variable model justification for principal component\n",
    "analysis, we are normally interested in inferring something about the\n",
    "latent variables given the data. This is the distribution, $$\n",
    "p(\\mathbf{ z}_{i, :} | \\mathbf{ y}_{i, :})\n",
    "$$ for any given data point. Determining this density turns out to be\n",
    "very similar to the approach for determining the Bayesian posterior of\n",
    "$\\mathbf{ w}$ in Bayesian linear regression, only this time we place the\n",
    "prior density over $\\mathbf{ z}_{i, :}$ instead of $\\mathbf{ w}$. The\n",
    "posterior is proportional to the joint density as follows, $$\n",
    "p(\\mathbf{ z}_{i, :} | \\mathbf{ y}_{i, :}) \\propto p(\\mathbf{ y}_{i,\n",
    ":}|\\mathbf{W}, \\mathbf{ z}_{i, :}, \\noiseStd^2) p(\\mathbf{ z}_{i, :})\n",
    "$$ And as in the Bayesian linear regression case we first consider the\n",
    "log posterior, $$\n",
    "\\log p(\\mathbf{ z}_{i, :} | \\mathbf{ y}_{i, :}) = \\log p(\\mathbf{ y}_{i, :}|\\mathbf{W},\n",
    "\\mathbf{ z}_{i, :}, \\noiseStd^2) + \\log p(\\mathbf{ z}_{i, :}) + \\text{const}\n",
    "$$ where the constant is not dependent on $\\mathbf{ z}$. As before we\n",
    "collect the quadratic terms in $\\mathbf{ z}_{i, :}$ and we assemble them\n",
    "into a Gaussian density over $\\mathbf{ z}$. $$\n",
    "\\log p(\\mathbf{ z}_{i, :} | \\mathbf{ y}_{i, :}) =\n",
    "-\\frac{1}{2\\noiseStd^2} (\\mathbf{ y}_{i, :} - \\mathbf{W}\\mathbf{ z}_{i,\n",
    ":})^\\top(\\mathbf{ y}_{i, :} - \\mathbf{W}\\mathbf{ z}_{i, :}) - \\frac{1}{2}\n",
    "\\mathbf{ z}_{i, :}^\\top \\mathbf{ z}_{i, :} + \\text{const}\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 0\n",
    "\n",
    "Multiply out the terms in the brackets. Then collect the quadratic term\n",
    "and the linear terms together. Show that the posterior has the form $$\n",
    "\\mathbf{ z}_{i, :} | \\mathbf{W}\\sim \\mathcal{N}\\left(\\boldsymbol{ \\mu}_x,\\mathbf{C}_x\\right)\n",
    "$$ where $$\n",
    "\\mathbf{C}_x = \\left(\\noiseStd^{-2}\n",
    "\\mathbf{W}^\\top\\mathbf{W}+ \\mathbf{I}\\right)^{-1}\n",
    "$$ and $$\n",
    "\\boldsymbol{ \\mu}_x\n",
    "= \\mathbf{C}_x \\noiseStd^{-2}\\mathbf{W}^\\top \\mathbf{ y}_{i, :} \n",
    "$$ Compare this to the posterior for the Bayesian linear regression from\n",
    "last week, do they have similar forms? What matches and what differs?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 0 Answer\n",
    "\n",
    "Write your answer to Exercise 0 here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Python Implementation of the Posterior\n",
    "--------------------------------------\n",
    "\n",
    "Now let’s implement the system in code."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 1\n",
    "\n",
    "Use the values for $\\mathbf{W}$ and $\\noiseStd^2$ you have computed,\n",
    "along with the data set $\\mathbf{Y}$ to compute the posterior density\n",
    "over $\\mathbf{Z}$. Write a function of the form"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your answer to Exercise 1 here\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "python mu\\_x, C\\_x = posterior(Y, W, sigma2)} where `mu_x` and `C_x` are\n",
    "the posterior mean and posterior covariance for the given $\\mathbf{Y}$.\n",
    "\n",
    "Don’t forget to subtract the mean of the data `Y` inside your function\n",
    "before computing the posterior: remember we assumed at the beginning of\n",
    "our analysis that the data had been centred (i.e. the mean was\n",
    "removed).}{20}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question 4 Answer Code\n",
    "# Write code for you answer to this question in this box\n",
    "# Do not delete these comments, otherwise you will get zero for this answer.\n",
    "# Make sure your code has run and the answer is correct *before* submitting your notebook for marking.\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "def posterior(Y, W, sigma2):\n",
    "    Y_cent = Y - Y.mean(0)\n",
    "    # Compute posterior over X\n",
    "    C_x = \n",
    "    mu_x = \n",
    "    return mu_x, C_x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Numerically Stable and Efficient Version\n",
    "----------------------------------------\n",
    "\n",
    "Just as we saw for and computation of a matrix such as\n",
    "$\\mathbf{Y}^\\top\\mathbf{Y}$ (or its centred version) can be a bad idea\n",
    "in terms of loss of numerical accuracy. Fortunately, we can find the\n",
    "eigenvalues and eigenvectors of the matrix $\\mathbf{Y}^\\top\\mathbf{Y}$\n",
    "without direct computation of the matrix. This can be done with the\n",
    "[*singular value\n",
    "decomposition*](http://en.wikipedia.org/wiki/Singular_value_decomposition).\n",
    "The singular value decompsition takes a matrix, $\\mathbf{Z}$ and\n",
    "represents it in the form, $$\n",
    "\\mathbf{Z} = \\mathbf{U}\\boldsymbol{\\Lambda}\\mathbf{V}^\\top\n",
    "$$ where $\\mathbf{U}$ is a matrix of orthogonal vectors in the columns,\n",
    "meaning $\\mathbf{U}^\\top\\mathbf{U} = \\mathbf{I}$. It has the same number\n",
    "of rows and columns as $\\mathbf{Z}$. The matrices $\\mathbf{\\Lambda}$ and\n",
    "$\\mathbf{V}$ are both square with dimensionality given by the number of\n",
    "columns of $\\mathbf{Z}$. The matrix $\\mathbf{\\Lambda}$ is *diagonal* and\n",
    "$\\mathbf{V}$ is an orthogonal matrix so\n",
    "$\\mathbf{V}^\\top\\mathbf{V} = \\mathbf{V}\\mathbf{V}^\\top = \\mathbf{I}$.\n",
    "The eigenvalues of the matrix $\\mathbf{Y}^\\top\\mathbf{Y}$ are then given\n",
    "by the singular values of the matrix $\\mathbf{Y}^\\top$ squared and the\n",
    "eigenvectors are given by $\\mathbf{U}$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Solution for $\\mathbf{W}$\n",
    "-------------------------\n",
    "\n",
    "Given the singular value decomposition of $\\mathbf{Y}$ then we have $$\n",
    "\\mathbf{W}=\n",
    "\\mathbf{U}\\mathbf{L}\\mathbf{R}^\\top\n",
    "$$ where $\\mathbf{R}$ is an arbitrary rotation matrix. This implies that\n",
    "the posterior is given by $$\n",
    "\\mathbf{C}_x =\n",
    "\\left[\\noiseStd^{-2}\\mathbf{R}\\mathbf{L}^2\\mathbf{R}^\\top + \\mathbf{I}\\right]^{-1}\n",
    "$$ because $\\mathbf{U}^\\top \\mathbf{U} = \\mathbf{I}$. Since, by\n",
    "convention, we normally take $\\mathbf{R} = \\mathbf{I}$ to ensure that\n",
    "the principal components are orthonormal we can write $$\n",
    "\\mathbf{C}_x = \\left[\\noiseStd^{-2}\\mathbf{L}^2 +\n",
    "\\mathbf{I}\\right]^{-1}\n",
    "$$ which implies that $\\mathbf{C}_x$ is actually diagonal with elements\n",
    "given by $$\n",
    "c_i = \\frac{\\noiseStd^2}{\\noiseStd^2 + \\ell^2_i}\n",
    "$$ and allows us to write $$\n",
    "\\boldsymbol{ \\mu}_x = [\\mathbf{L}^2 + \\noiseStd^2\n",
    "\\mathbf{I}]^{-1} \\mathbf{L} \\mathbf{U}^\\top \\mathbf{ y}_{i, :}\n",
    "$$ $$\n",
    "\\boldsymbol{ \\mu}_x = \\mathbf{D}\\mathbf{U}^\\top \\mathbf{ y}_{i, :}\n",
    "$$ where $\\mathbf{D}$ is a diagonal matrix with diagonal elements given\n",
    "by $d_{i} = \\frac{\\ell_i}{\\noiseStd^2 + \\ell_i^2}$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy as sp\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# probabilistic PCA algorithm using SVD\n",
    "def ppca(Y, q, center=True):\n",
    "    \"\"\"Probabilistic PCA through singular value decomposition\"\"\"\n",
    "    # remove mean\n",
    "    if center:\n",
    "        Y_cent = Y - Y.mean(0)\n",
    "    else:\n",
    "        Y_cent = Y\n",
    "        \n",
    "    # Comute singluar values, discard 'R' as we will assume orthogonal\n",
    "    U, sqlambd, _ = sp.linalg.svd(Y_cent.T,full_matrices=False)\n",
    "    lambd = (sqlambd**2)/Y.shape[0]\n",
    "    # Compute residual and extract eigenvectors\n",
    "    sigma2 = np.sum(lambd[q:])/(Y.shape[1]-q)\n",
    "    ell = np.sqrt(lambd[:q]-sigma2)\n",
    "    return U[:, :q], ell, sigma2\n",
    "\n",
    "def posterior(Y, U, ell, sigma2, center=True):\n",
    "    \"\"\"Posterior computation for the latent variables given the eigendecomposition.\"\"\"\n",
    "    if center:\n",
    "        Y_cent = Y - Y.mean(0)\n",
    "    else:\n",
    "        Y_cent = Y\n",
    "    C_x = np.diag(sigma2/(sigma2+ell**2))\n",
    "    d = ell/(sigma2+ell**2)\n",
    "    mu_x = np.dot(Y_cent, U)*d[None, :]\n",
    "    return mu_x, C_x"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Difficulty for Probabilistic Approaches\n",
    "---------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import teaching_plots as plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot.non_linear_difficulty_plot_3(diagrams='./dimred/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The challenge for composition of probabilistic models is that you need\n",
    "to propagate a probability densities through non linear mappings. This\n",
    "allows you to create broader classes of probability density.\n",
    "Unfortunately it renders the resulting densities *intractable*.\n",
    "\n",
    "<img src=\"http://inverseprobability.com/talks/slides/../slides/diagrams/dimred/nonlinear-mapping-3d-plot.svg\" class=\"\" width=\"80%\" style=\"vertical-align:middle;\">\n",
    "\n",
    "Figure: <i>A two dimensional grid mapped into three dimensions to form a\n",
    "two dimensional manifold.</i>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot.non_linear_difficulty_plot_2(diagrams='./dimred/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"http://inverseprobability.com/talks/slides/../slides/diagrams/dimred/nonlinear-mapping-2d-plot.svg\" class=\"\" width=\"80%\" style=\"vertical-align:middle;\">\n",
    "\n",
    "Figure: <i>A one dimensional line mapped into two dimensions by two\n",
    "separate independent functions. Each point can be mapped exactly through\n",
    "the mappings.</i>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot.non_linear_difficulty_plot_1(diagrams='./dimred')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"http://inverseprobability.com/talks/slides/../slides/diagrams/dimred/gaussian-through-nonlinear.svg\" class=\"\" width=\"100%\" style=\"vertical-align:middle;\">\n",
    "\n",
    "Figure: <i>A Gaussian density over the input of a non linear function\n",
    "leads to a very non Gaussian output. Here the output is multimodal.</i>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Getting Started and Downloading Data\n",
    "------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import GPy\n",
    "import string"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following code is for plotting and to prepare the bigger models for\n",
    "later useage. If you are interested, you can have a look, but this is\n",
    "not essential."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "import teaching_plots as plot\n",
    "import mlai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = [\"#3FCC94\", \"#DD4F23\", \"#C6D63B\", \"#D44271\", \n",
    "          \"#E4A42C\", \"#4F9139\", \"#6DDA4C\", \"#85831F\", \n",
    "          \"#B36A29\", \"#CF4E4A\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_model(X, which_dims, labels):\n",
    "    fig, ax = plt.subplots(figsize=plot.big_figsize)\n",
    "    X = X[:,which_dims]\n",
    "    ulabs = []\n",
    "    for lab in labels:\n",
    "        if not lab in ulabs:\n",
    "            ulabs.append(lab)\n",
    "            pass\n",
    "        pass\n",
    "    for i, lab in enumerate(ulabs):\n",
    "        ax.scatter(*X[labels==lab].T,marker='o',color=colors[i],label=lab)\n",
    "        pass\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this lab, we’ll use a data set containing all handwritten digits\n",
    "from $0 \\cdots 9$ handwritten, provided by de Campos et al. (2009). We\n",
    "will only use some of the digits for the demonstrations in this lab\n",
    "class, but you can edit the code below to select different subsets of\n",
    "the digit data as you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "which = [0,1,2,6,7,9] # which digits to work on\n",
    "data = pods.datasets.decampos_digits(which_digits=which)\n",
    "Y = data['Y']\n",
    "labels = data['str_lbls']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can try to plot some of the digits using `plt.matshow` (the digit\n",
    "images have size `16x16`)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Principal Component Analysis\n",
    "----------------------------\n",
    "\n",
    "Principal component analysis (PCA) finds a rotation of the observed\n",
    "outputs, such that the rotated principal component (PC) space maximizes\n",
    "the variance of the data observed, sorted from most to least important\n",
    "(most to least variable in the corresponding PC).\n",
    "\n",
    "In order to apply PCA in an easy way, we have included a PCA module in\n",
    "pca.py. You can import the module by import \\<path.to.pca\\> (without the\n",
    "ending .py!). To run PCA on the digits we have to reshape (Hint:\n",
    "np.reshape ) digits .\n",
    "\n",
    "-   What is the right shape $n\\times p$ to use?\n",
    "\n",
    "We will call the reshaped observed outputs $\\mathbf{Y}$ in the\n",
    "following."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Yn = Y#Y-Y.mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let’s run PCA on the reshaped dataset $\\mathbf{Y}$:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from GPy.util import pca\n",
    "p = pca.pca(Y) # create PCA class with digits dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The resulting plot will show the lower dimensional representation of the\n",
    "digits in 2 dimensions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p.plot_fracs(20) # plot first 20 eigenvalue fractions\n",
    "p.plot_2d(Y,labels=labels.flatten(), colors=colors)\n",
    "pb.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Gaussian Process Latent Variable Model\n",
    "--------------------------------------\n",
    "\n",
    "The Gaussian Process Latent Variable Model (GP-LVM) (Lawrence, 2005)\n",
    "embeds PCA into a Gaussian process framework, where the latent inputs\n",
    "$\\mathbf{Z}$ are learnt as hyperparameters and the mapping variables\n",
    "$\\mathbf{W}$ are integrated out. The advantage of this interpretation is\n",
    "it allows PCA to be generalized in a non linear way by replacing the\n",
    "resulting *linear* covariance witha non linear covariance. But first,\n",
    "let’s see how GPLVM is equivalent to PCA using an automatic relevance\n",
    "determination (ARD, see e.g. Bishop (2006)) linear kernel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dim = 4 # How many latent dimensions to use\n",
    "kernel = GPy.kern.Linear(input_dim, ARD=True) # ARD kernel\n",
    "m = GPy.models.GPLVM(Yn, input_dim=input_dim, kernel=kernel)\n",
    "\n",
    "m.optimize(messages=1, max_iters=1000) # optimize for 1000 iterations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m.kern.plot_ARD()\n",
    "plot_model(m.X, m.linear.variances.argsort()[-2:], labels.flatten())\n",
    "pb.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can see the solution with a linear kernel is the same as the PCA\n",
    "solution with the exception of rotational changes and axis flips.\n",
    "\n",
    "For the sake of time, the solution you see was only running for 1000\n",
    "iterations, thus it might not be converged fully yet. The GP-LVM\n",
    "proceeds by iterative optimization of the *inputs* to the covariance. As\n",
    "we saw in the lecture earlier, for the linear covariance, these latent\n",
    "points can be optimized with an eigenvalue problem, but generally, for\n",
    "non-linear covariance functions, we are obliged to use gradient based\n",
    "optimization."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CMU Mocap Database\n",
    "------------------\n",
    "\n",
    "Motion capture data from the CMU motion capture data base (CMU Motion\n",
    "Capture Lab, 2003)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pods"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can download any subject and motion from the data set. Here we will\n",
    "download motion `01` from subject `35`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subject='35' \n",
    "motion=['01']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pods.datasets.cmu_mocap(subject, motion)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data dictionary contains the keys ‘Y’ and ‘skel’, which represent\n",
    "the data and the skeleton.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['Y'].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data was used in the hierarchical GP-LVM paper (Lawrence and Moore,\n",
    "2007) in an experiment that was also recreated in the Deep Gaussian\n",
    "process paper (Damianou and Lawrence, 2013)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(data['citation'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And extra information about the data is included, as standard, under the\n",
    "keys `info` and `details`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(data['info'])\n",
    "print()\n",
    "print(data['details'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The original data has the figure moving across the floor during the\n",
    "motion capture sequence. We can make the figure walk ‘in place’, by\n",
    "setting the x, y, z positions of the root node to zero. This makes it\n",
    "easier to visualize the result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make figure move in place.\n",
    "data['Y'][:, 0:3] = 0.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also remove the mean of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Y = data['Y']\n",
    "Y_mean = Y.mean(0)\n",
    "Y_std = Y.std(0)\n",
    "Yhat = (Y-Y_mean)/Y_std"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we create the GP-LVM model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import GPy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = GPy.models.GPLVM(Yhat, 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we optimize the model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.optimize(messages=True, max_f_eval=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import teaching_plots as plot\n",
    "import mlai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = model.plot_latent()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y = model.Y[0, :]\n",
    "data_show = GPy.plotting.matplot_dep.visualize.skeleton_show(y[np.newaxis, :], \n",
    "                                                             data['skel'])\n",
    "lvm_visualizer = GPy.plotting.matplot_dep.visualize.lvm(model.X[0].copy(), \n",
    "                                                        model, \n",
    "                                                        data_show, \n",
    "                                                        latent_axes=ax)\n",
    "input('Press enter to finish')\n",
    "lvm_visualizer.close()\n",
    "data_show.close()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Example: Latent Doodle Space\n",
    "----------------------------\n",
    "\n",
    "[![](attachment:https://i.vimeocdn.com/video/\\id_640x480.jpg)](https://vimeo.com/3235882#t=)\n",
    "\n",
    "Figure: <i>The latent doodle space idea of Baxter and Anjyo (2006)\n",
    "manages to build a smooth mapping across very sparse data.</i>\n",
    "\n",
    "**Generalization with much less Data than Dimensions**\n",
    "\n",
    "-   Powerful uncertainly handling of GPs leads to surprising properties.\n",
    "\n",
    "-   Non-linear models can be used where there are fewer data points than\n",
    "    dimensions *without overfitting*.\n",
    "\n",
    "<span style=\"text-align:right\">(Baxter and Anjyo, 2006)</span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Example: Continuous Character Control\n",
    "-------------------------------------\n",
    "\n",
    "(Levine et al., 2012){style=“text-align:right”} - Graph diffusion prior\n",
    "for enforcing connectivity between motions.\n",
    "$$\\log p(\\mathbf{X}) = w_c \\sum_{i,j} \\log K_{ij}^d$$ with the graph\n",
    "diffusion kernel $\\mathbf{K}^d$ obtain from\n",
    "$$K_{ij}^d = \\exp(\\beta \\mathbf{H})\n",
    "    \\qquad \\text{with} \\qquad \\mathbf{H} = -\\mathbf{T}^{-1/2} \\mathbf{L} \\mathbf{T}^{-1/2}$$\n",
    "the graph Laplacian, and $\\mathbf{T}$ is a diagonal matrix with\n",
    "$T_{ii} = \\sum_j w(\\mathbf{ x}_i, \\mathbf{ x}_j)$,\n",
    "$$L_{ij} = \\begin{cases} \\sum_k w(\\mathbf{ x}_i,\\mathbf{ x}_k) & \\text{if $i=j$}\n",
    "    \\\\\n",
    "    -w(\\mathbf{ x}_i,\\mathbf{ x}_j) &\\text{otherwise.}\n",
    "    \\end{cases}$$ and\n",
    "$w(\\mathbf{ x}_i,\\mathbf{ x}_j) = || \\mathbf{ x}_i - \\mathbf{ x}_j||^{-p}$\n",
    "measures similarity."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Character Control: Results\n",
    "--------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.lib.display import YouTubeVideo\n",
    "YouTubeVideo('hr3pdDl5IAg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Figure: <i>Character control in the latent space described the the\n",
    "GP-LVM Levine et al. (2012)</i>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data for Blastocyst Development in Mice: Single Cell TaqMan Arrays\n",
    "------------------------------------------------------------------\n",
    "\n",
    "Now we analyze some single cell data from Guo et al. (2010). Tey\n",
    "performed qPCR TaqMan array on single cells from the developing\n",
    "blastocyst in mouse. The data is taken from the early stages of\n",
    "development when the Blastocyst is forming. At the 32 cell stage the\n",
    "data is already separated into the trophectoderm (TE) which goes onto\n",
    "form the placenta and the inner cellular mass (ICM). The ICM further\n",
    "differentiates into the epiblast (EPI)—which gives rise to the endoderm,\n",
    "mesoderm and ectoderm—and the primitive endoderm (PE) which develops\n",
    "into the amniotic sack. Guo et al. (2010) selected 48 genes for\n",
    "expression measurement. They labelled the resulting cells and their\n",
    "labels are included as an aide to visualization.\n",
    "\n",
    "They first visualized their data using principal component analysis. In\n",
    "the first two principal components this fails to separate the domains.\n",
    "This is perhaps because the principal components are dominated by the\n",
    "variation in the 64 cell systems. This in turn may be because there are\n",
    "more cells from the data set in that regime, and may be because the\n",
    "natural variation is greater. We first recreate their visualization\n",
    "using principal component analysis.\n",
    "\n",
    "In this notebook we will perform PCA on the original data, showing that\n",
    "the different regimes do not separate.\n",
    "\n",
    "Next we load in the data. We’ve provided a convenience function for\n",
    "loading in the data with `pods`. It is loaded in as a `pandas`\n",
    "DataFrame. This allows us to summarize it with the `describe` attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pods.datasets.singlecell()\n",
    "Y = data['Y']\n",
    "Y.describe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Principal Component Analysis\n",
    "----------------------------\n",
    "\n",
    "Now we follow Guo et al. (2010) in performing PCA on the data. Rather\n",
    "than calling a ‘PCA routine’, here we break the algorithm down into its\n",
    "steps: compute the data covariance, compute the eigenvalues and\n",
    "eigenvectors and sort according to magnitude of eigenvalue. Because we\n",
    "want to visualize the data, we’ve chose to compute the eigenvectors of\n",
    "the *inner product matrix* rather than the covariance matrix. This\n",
    "allows us to plot the eigenvalues directly. However, this is less\n",
    "efficient (in this case because the number of genes is smaller than the\n",
    "number of data) than computing the eigendecomposition of the covariance\n",
    "matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain a centred version of data.\n",
    "centredY = Y - Y.mean()\n",
    "# compute inner product matrix\n",
    "C = np.dot(centredY,centredY.T)\n",
    "# perform eigendecomposition\n",
    "V, U = np.linalg.eig(C)\n",
    "# sort eigenvalues and vectors according to size\n",
    "ind = V.argsort()\n",
    "ev = V[ind[::-1]]\n",
    "U = U[:, ind[::-1]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To visualize the result, we now construct a simple helper function. This\n",
    "will ensure that the plots have the same legends as the GP-LVM plots we\n",
    "use below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import GPy\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_labels(ax, x, y, labels, symbols):\n",
    "    \"\"\"A small helper function for plotting with labels\"\"\"\n",
    "    # make sure labels are in order of input:\n",
    "    ulabels = []\n",
    "    for lab in labels:\n",
    "        if not lab in ulabels:\n",
    "            ulabels.append(lab)\n",
    "    for i, label in enumerate(ulabels):\n",
    "        symbol = symbols[i % len(symbols)]\n",
    "        ind = labels == label\n",
    "        ax.plot(x[ind], y[ind], symbol)\n",
    "    ax.legend(ulabels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PCA Result\n",
    "----------\n",
    "\n",
    "Now, using the helper function we can plot the results with appropriate\n",
    "labels."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import teaching_plots as plot\n",
    "import mlai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=plot.big_figsize)\n",
    "plot_labels(ax=ax, U[:, 0], U[:, 1], data['labels'], '<>^vsd')\n",
    "\n",
    "mlai.write_figure('singlecell-data-pca.svg', directory='./datasets')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"http://inverseprobability.com/talks/slides/../slides/diagrams/datasets/singlecell-data-pca.svg\" class=\"\" width=\"60%\" style=\"vertical-align:middle;\">\n",
    "\n",
    "Figure: <i>First two principal compoents of the Guo et al. (2010)\n",
    "blastocyst development data.</i>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "GP-LVM on the Data\n",
    "------------------\n",
    "\n",
    "Work done as a collaboration between Max Zwiessele, Oliver Stegle and\n",
    "Neil D. Lawrence.\n",
    "\n",
    "Then, we follow Buettner and Theis (2012) in applying the GP-LVM to the\n",
    "data. There is a slight pathology in the result, one which they fixed by\n",
    "using priors that were dependent on the developmental stage. We then\n",
    "show how the Bayesian GP-LVM doesn’t exhibit those pathologies and gives\n",
    "a nice results that seems to show the lineage of the cells.\n",
    "\n",
    "They used modified prior to ensure that small differences between cells\n",
    "at the same differential stage were preserved. Here we apply a standard\n",
    "GP-LVM (no modified prior) to the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import GPy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kernel = GPy.kern.RBF(2)+GPy.kern.Bias(2)\n",
    "model = GPy.models.GPLVM(Y.values, 2, kernel=kernel)\n",
    "model.optimize(messages=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import teaching_plots as plot\n",
    "import mlai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=plot.big_figsize)\n",
    "model.plot_latent(ax=ax, labels=data['labels'], marker='<>^vsd')\n",
    "\n",
    "mlai.write_figure('singlecell-gplvm.svg', directory='./gplvm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"http://inverseprobability.com/talks/slides/../slides/diagrams/gplvm/singlecell-gplvm.svg\" class=\"\" width=\"60%\" style=\"vertical-align:middle;\">\n",
    "\n",
    "Figure: <i>Visualisation of the Guo et al. (2010) blastocyst development\n",
    "data with the GP-LVM.</i>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=plot.big_wide_figsize)\n",
    "model.kern.plot_ARD(ax=ax)\n",
    "\n",
    "mlai.write_figure('singlecell-gplvm-ard.svg', directory='./gplvm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"http://inverseprobability.com/talks/slides/../slides/diagrams/gplvm/singlecell-gplvm-ard.svg\" class=\"\" width=\"80%\" style=\"vertical-align:middle;\">\n",
    "\n",
    "Figure: <i>The ARD parameters of the GP-LVM for the Guo et al. (2010)\n",
    "blastocyst development data.</i>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Blastocyst Data: Isomap\n",
    "-----------------------\n",
    "\n",
    "Isomap first builds a neighbourhood graph, and then uses distances along\n",
    "this graph to approximate the geodesic distance between points. These\n",
    "distances are then visualized by performing classical multidimensional\n",
    "scaling (which involves computing the eigendecomposition of the centred\n",
    "distance matrix). As the neighborhood size is increased to match the\n",
    "data, principal component analysis is recovered (or strictly speaking,\n",
    "principal coordinate analysis). The fewer the neighbors, the more\n",
    "‘non-linear’ the isomap embeddings are."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sklearn.manifold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = 10\n",
    "model = sklearn.manifold.Isomap(n_neighbors=n_neighbors, n_components=2)\n",
    "X = model.fit_transform(Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import teaching_plots as plot\n",
    "import mlai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=plot.big_figsize)\n",
    "plot_labels(ax, X[:, 0], X[:, 1], data['labels'], '<>^vsd')\n",
    "\n",
    "\n",
    "mlai.write_figure('singlecell-isomap.svg', directory='./dimred')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"http://inverseprobability.com/talks/slides/../slides/diagrams/dimred/singlecell-isomap.svg\" class=\"\" width=\"60%\" style=\"vertical-align:middle;\">\n",
    "\n",
    "Figure: <i>Visualisation of the Guo et al. (2010) blastocyst development\n",
    "data with Isomap.</i>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Blastocyst Data: Locally Linear Embedding\n",
    "-----------------------------------------\n",
    "\n",
    "Next we try locally linear embedding. In locally linear embedding a\n",
    "neighborhood is also computed. Each point is then reconstructed by it’s\n",
    "neighbors using a linear weighting. This implies a locally linear patch\n",
    "is being fitted to the data in that region. These patches are\n",
    "assimilated into a large $n\\times n$ matrix and a lower dimensional data\n",
    "set which reflects the same relationships is then sought. Quite a large\n",
    "number of neighbours needs to be selected for the data to not collapse\n",
    "in on itself. When a large number of neighbours is selected the\n",
    "embedding is more linear and begins to look like PCA. However, the\n",
    "algorithm does *not* converge to PCA in the limit as the number of\n",
    "neighbors approaches $n$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sklearn.manifold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_neighbors = 50\n",
    "model = sklearn.manifold.LocallyLinearEmbedding(n_neighbors=n_neighbors, n_components=2)\n",
    "X = model.fit_transform(Y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import teaching_plots as plot\n",
    "import mlai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=plot.big_figsize)\n",
    "model.plot_latent(ax=ax, X[:, 0], X[:, 1], data['labels'], '<>^vsd')\n",
    "\n",
    "mlai.write_figure('singlecell-lle.svg', directory='./dimred')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"http://inverseprobability.com/talks/slides/../slides/diagrams/dimred/singlecell-lle.svg\" class=\"\" width=\"60%\" style=\"vertical-align:middle;\">\n",
    "\n",
    "Figure: <i>Visualisation of the Guo et al. (2010) blastocyst development\n",
    "data with a locally linear embedding.</i>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Thanks!\n",
    "-------\n",
    "\n",
    "For more information on these subjects and more you might want to check\n",
    "the following resources.\n",
    "\n",
    "-   twitter: [@lawrennd](https://twitter.com/lawrennd)\n",
    "-   podcast: [The Talking Machines](http://thetalkingmachines.com)\n",
    "-   newspaper: [Guardian Profile\n",
    "    Page](http://www.theguardian.com/profile/neil-lawrence)\n",
    "-   blog:\n",
    "    [http://inverseprobability.com](http://inverseprobability.com/blog.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "References\n",
    "----------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Baxter, W.V., Anjyo, K.-I., 2006. Latent doodle space, in: EUROGRAPHICS.\n",
    "Vienna, Austria, pp. 477–485.\n",
    "<https://doi.org/10.1111/j.1467-8659.2006.00967.x>\n",
    "\n",
    "Bishop, C.M., 2006. Pattern recognition and machine learning. springer.\n",
    "\n",
    "Buettner, F., Theis, F.J., 2012. A novel approach for resolving\n",
    "differences in single-cell gene expression patterns from zygote to\n",
    "blastocyst. Bioinformatics 28, i626–i632.\n",
    "<https://doi.org/10.1093/bioinformatics/bts385>\n",
    "\n",
    "CMU Motion Capture Lab, 2003. The cmu mocap database.\n",
    "\n",
    "Damianou, A., Lawrence, N.D., 2013. Deep Gaussian processes, in:. pp.\n",
    "207–215.\n",
    "\n",
    "de Campos, T.E., Babu, B.R., Varma, M., 2009. Character recognition in\n",
    "natural images, in: Proceedings of the Fourth International Conference\n",
    "on Computer Vision Theory and Applications - Volume 2: VISAPP,\n",
    "(Visigrapp 2009). INSTICC; SciTePress, pp. 273–280.\n",
    "<https://doi.org/10.5220/0001770102730280>\n",
    "\n",
    "Guo, G., Huss, M., Tong, G.Q., Wang, C., Sun, L.L., Clarke, N.D.,\n",
    "Robsonemail, P., 2010. Resolution of cell fate decisions revealed by\n",
    "single-cell gene expression analysis from zygote to blastocyst.\n",
    "Developmental Cell 18, 675–685.\n",
    "<https://doi.org/10.1016/j.devcel.2010.02.012>\n",
    "\n",
    "Lawrence, N.D., 2005. Probabilistic non-linear principal component\n",
    "analysis with Gaussian process latent variable models. Journal of\n",
    "Machine Learning Research 6, 1783–1816.\n",
    "\n",
    "Lawrence, N.D., Moore, A.J., 2007. Hierarchical Gaussian process latent\n",
    "variable models, in:. pp. 481–488.\n",
    "\n",
    "Levine, S., Wang, J.M., Haraux, A., Popović, Z., Koltun, V., 2012.\n",
    "Continuous character control with low-dimensional embeddings. ACM\n",
    "Transactions on Graphics (SIGGRAPH 2012) 31.\n",
    "\n",
    "Roweis, S.T., n.d. EM algorithms for PCA and SPCA, in:. pp. 626–632.\n",
    "\n",
    "Tipping, M.E., Bishop, C.M., 1999. Mixtures of probabilistic principal\n",
    "component analysers. Neural Computation 11, 443–482."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
